--- ./vio/Makefile.am	2013-08-17 04:51:01.000000000 +0400
+++ ../x1/mysql-5.1.65/./vio/Makefile.am	2012-07-12 13:12:07.000000000 +0400
@@ -21,7 +21,6 @@
 noinst_HEADERS =	vio_priv.h
 
 libvio_a_SOURCES =	vio.c viosocket.c viossl.c viosslfactories.c
-libvio_a_CFLAGS = -fPIC
 
 EXTRA_DIST=		CMakeLists.txt
 
--- ./sql/Makefile.am	2013-08-24 01:32:44.000000000 +0400
+++ ../x1/mysql-5.1.65/./sql/Makefile.am	2012-07-12 13:12:07.000000000 +0400
@@ -25,26 +25,25 @@
 			-I$(top_srcdir)/regex -I$(srcdir) $(openssl_includes)
 WRAPLIBS=		@WRAPLIBS@
 SUBDIRS =		share
+libexec_PROGRAMS =	mysqld
 EXTRA_PROGRAMS =	gen_lex_hash
 bin_PROGRAMS =		mysql_tzinfo_to_sql
 
-noinst_LTLIBRARIES=	libndb.la udf_example.la libmysqldlib.la
+noinst_LTLIBRARIES=	libndb.la \
+			udf_example.la
 
 SUPPORTING_LIBS =	$(top_builddir)/vio/libvio.a \
 			$(top_builddir)/mysys/libmysys.a \
 			$(top_builddir)/dbug/libdbug.a \
 			$(top_builddir)/regex/libregex.a \
 			$(top_builddir)/strings/libmystrings.a
-libmysqldlib_la_DEPENDENCIES=	@mysql_plugin_libs@ $(SUPPORTING_LIBS) libndb.la
+mysqld_DEPENDENCIES=	@mysql_plugin_libs@ $(SUPPORTING_LIBS) libndb.la
 LDADD = $(SUPPORTING_LIBS) @ZLIB_LIBS@ @NDB_SCI_LIBS@
-libmysqldlib_la_LIBADD =		libndb.la \
+mysqld_LDADD =		libndb.la \
 			@MYSQLD_EXTRA_LDFLAGS@ \
 			@mysql_plugin_libs@ \
 			$(LDADD)  $(CXXLDFLAGS) $(WRAPLIBS) @LIBDL@ \
 			$(yassl_libs) $(openssl_libs) @MYSQLD_EXTRA_LIBS@
-libmysqldlib_la_LDFLAGS = -fPIC -module -rpath $(pkglibdir)
-libmysqldlib_la_CXXFLAGS = -fPIC 
-libmysqldlib_la_CFLAGS = -fPIC 
 
 noinst_HEADERS =	item.h item_func.h item_sum.h item_cmpfunc.h \
 			item_strfunc.h item_timefunc.h \
@@ -80,7 +79,7 @@
 			sql_partition.h partition_info.h partition_element.h \
 			contributors.h sql_servers.h
 
-libmysqldlib_la_SOURCES = sql_lex.cc sql_handler.cc sql_partition.cc \
+mysqld_SOURCES =	sql_lex.cc sql_handler.cc sql_partition.cc \
 			item.cc item_sum.cc item_buff.cc item_func.cc \
 			item_cmpfunc.cc item_strfunc.cc item_timefunc.cc \
 			thr_malloc.cc item_create.cc item_subselect.cc \
@@ -89,7 +88,7 @@
 			net_serv.cc protocol.cc sql_state.c \
 			lock.cc my_lock.c \
 			sql_string.cc sql_manager.cc sql_map.cc \
-			mysqldlib.cc password.c hash_filo.cc hostname.cc \
+			mysqld.cc password.c hash_filo.cc hostname.cc \
 			sql_connect.cc scheduler.cc sql_parse.cc \
 			set_var.cc sql_yacc.yy \
 			sql_base.cc table.cc sql_select.cc sql_insert.cc \
@@ -126,8 +125,7 @@
 			sql_builtin.cc sql_tablespace.cc partition_info.cc \
 			sql_servers.cc event_parse_data.cc
 
-
-nodist_libmysqldlib_la_SOURCES =	mini_client_errors.c pack.c client.c my_time.c my_user.c 
+nodist_mysqld_SOURCES =	mini_client_errors.c pack.c client.c my_time.c my_user.c 
 
 libndb_la_CPPFLAGS=	@ndbcluster_includes@
 libndb_la_SOURCES=	ha_ndbcluster.cc \
--- ./strings/Makefile.am	2013-08-17 05:07:03.000000000 +0400
+++ ../x1/mysql-5.1.65/./strings/Makefile.am	2012-07-12 13:12:07.000000000 +0400
@@ -38,7 +38,6 @@
 endif
 
 libmystrings_a_SOURCES = $(ASRCS) $(CSRCS)
-libmystrings_a_CFLAGS = -fPIC 
 noinst_PROGRAMS = conf_to_src
 CLEANFILES = str_test uctypedump test_decimal
 # Default charset definitions
--- ./storage/myisam/Makefile.am	2013-08-24 01:33:52.000000000 +0400
+++ ../x1/mysql-5.1.65/./storage/myisam/Makefile.am	2012-07-12 13:12:07.000000000 +0400
@@ -31,7 +31,6 @@
 pkgdata_DATA =		mi_test_all mi_test_all.res
 
 pkglib_LIBRARIES =	libmyisam.a
-noinst_LTLIBRARIES=	libmyisamreader.la
 bin_PROGRAMS =		myisamchk myisamlog myisampack myisam_ftdump
 myisamchk_DEPENDENCIES=	$(LIBRARIES)
 myisamchk_LDADD=		@CLIENT_EXTRA_LDFLAGS@ libmyisam.a \
@@ -48,14 +47,6 @@
 			$(top_builddir)/mysys/libmysys.a \
 			$(top_builddir)/dbug/libdbug.a \
 			$(top_builddir)/strings/libmystrings.a @ZLIB_LIBS@
-libmyisamreader_la_DEPENDENCIES=$(LIBRARIES)
-libmyisamreader_la_LIBADD =		@CLIENT_EXTRA_LDFLAGS@ libmyisam.a \
-			$(top_builddir)/mysys/libmysys.a \
-			$(top_builddir)/dbug/libdbug.a \
-			$(top_builddir)/strings/libmystrings.a @ZLIB_LIBS@
-libmyisamreader_la_SOURCES = myisamreader.c
-libmyisamreader_la_LDFLAGS = -fPIC -module -rpath $(pkglibdir)
-libmyisamreader_la_CFLAGS = -fPIC
 noinst_PROGRAMS =	mi_test1 mi_test2 mi_test3 rt_test sp_test #ft_test1 ft_eval
 noinst_HEADERS =	myisamdef.h rt_index.h rt_key.h rt_mbr.h sp_defs.h \
 			fulltext.h ftdefs.h ft_test1.h ft_eval.h \
@@ -109,8 +100,6 @@
 			ft_update.c ft_boolean_search.c ft_nlq_search.c sort.c \
 			ha_myisam.cc \
 			rt_index.c rt_key.c rt_mbr.c rt_split.c sp_key.c
-libmyisam_a_CXXFLAGS = -fPIC
-libmyisam_a_CFLAGS = -fPIC
 CLEANFILES =		test?.MY? FT?.MY? isam.log mi_test_all rt_test.MY? sp_test.MY?
 
 # Move to automake rules ?
--- ./storage/myisammrg/Makefile.am	2013-08-17 04:50:14.000000000 +0400
+++ ../x1/mysql-5.1.65/./storage/myisammrg/Makefile.am	2012-07-12 13:12:07.000000000 +0400
@@ -36,8 +36,6 @@
 			myrg_rprev.c myrg_queue.c myrg_write.c myrg_range.c \
 			ha_myisammrg.cc \
 			myrg_rnext_same.c myrg_records.c
-libmyisammrg_a_CXXFLAGS = -fPIC -DPIC
-libmyisammrg_a_CFLAGS = -fPIC -DPIC
 
 
 EXTRA_DIST =		CMakeLists.txt plug.in
--- ./storage/heap/Makefile.am	2013-08-17 04:45:29.000000000 +0400
+++ ../x1/mysql-5.1.65/./storage/heap/Makefile.am	2012-07-12 13:12:07.000000000 +0400
@@ -34,13 +34,11 @@
 			$(top_builddir)/mysys/libmysys.a \
 			$(top_builddir)/dbug/libdbug.a \
 			$(top_builddir)/strings/libmystrings.a
-hp_test1_CFLAGS = -fPIC
 hp_test2_LDFLAGS = @NOINST_LDFLAGS@
 hp_test2_LDADD =	libheap.a \
 			$(top_builddir)/mysys/libmysys.a \
 			$(top_builddir)/dbug/libdbug.a \
 			$(top_builddir)/strings/libmystrings.a
-hp_test2_CFLAGS = -fPIC
 noinst_HEADERS =	heapdef.h ha_heap.h
 libheap_a_SOURCES =	hp_open.c hp_extra.c hp_close.c hp_panic.c hp_info.c \
 			hp_rrnd.c hp_scan.c hp_update.c hp_write.c hp_delete.c \
@@ -49,8 +47,6 @@
 			hp_rkey.c hp_block.c \
 			ha_heap.cc \
 			hp_hash.c _check.c _rectest.c hp_static.c
-libheap_a_CFLAGS = -fPIC -DPIC
-libheap_a_CXXFLAGS = -fPIC -DPIC
 
 
 EXTRA_DIST =		CMakeLists.txt plug.in
--- ./storage/csv/Makefile.am	2013-08-17 04:37:12.000000000 +0400
+++ ../x1/mysql-5.1.65/./storage/csv/Makefile.am	2012-07-12 13:12:07.000000000 +0400
@@ -32,12 +32,12 @@
 EXTRA_LTLIBRARIES =	ha_csv.la
 pkglib_LTLIBRARIES =	@plugin_csv_shared_target@
 ha_csv_la_LDFLAGS =	-module -rpath $(MYSQLLIBdir)
-ha_csv_la_CXXFLAGS =	$(AM_CXXFLAGS) -DMYSQL_PLUGIN -fPIC
+ha_csv_la_CXXFLAGS =	$(AM_CXXFLAGS) -DMYSQL_PLUGIN
 ha_csv_la_SOURCES =	transparent_file.cc ha_tina.cc 
 
 EXTRA_LIBRARIES =	libcsv.a
 noinst_LIBRARIES =	@plugin_csv_static_target@
-libcsv_a_CXXFLAGS =	$(AM_CXXFLAGS) -fPIC
+libcsv_a_CXXFLAGS =	$(AM_CXXFLAGS)
 libcsv_a_SOURCES =	transparent_file.cc ha_tina.cc
 
 EXTRA_DIST =		CMakeLists.txt plug.in
--- ./regex/Makefile.am	2013-08-17 05:06:28.000000000 +0400
+++ ../x1/mysql-5.1.65/./regex/Makefile.am	2012-07-12 13:12:07.000000000 +0400
@@ -20,7 +20,6 @@
 LDADD=			libregex.a $(top_builddir)/strings/libmystrings.a
 noinst_HEADERS =	cclass.h cname.h regex2.h utils.h engine.c my_regex.h
 libregex_a_SOURCES =	regerror.c regcomp.c regexec.c regfree.c reginit.c
-libregex_a_CFLAGS = -fPIC
 noinst_PROGRAMS =	re
 re_SOURCES =		split.c debug.c main.c
 re_LDFLAGS= @NOINST_LDFLAGS@
--- ./dbug/Makefile.am	2013-08-17 05:05:54.000000000 +0400
+++ ../x1/mysql-5.1.65/./dbug/Makefile.am	2012-07-12 13:12:07.000000000 +0400
@@ -20,7 +20,6 @@
 pkglib_LIBRARIES =      libdbug.a
 noinst_HEADERS =        dbug_long.h
 libdbug_a_SOURCES =     dbug.c sanity.c
-libdbug_a_CFLAGS = -fPIC
 EXTRA_DIST =            CMakeLists.txt example1.c example2.c example3.c \
                         user.r monty.doc dbug_add_tags.pl \
                         my_main.c main.c factorial.c dbug_analyze.c \
--- ./mysys/Makefile.am	2013-08-17 05:13:32.000000000 +0400
+++ ../x1/mysql-5.1.65/./mysys/Makefile.am	2012-07-12 13:12:07.000000000 +0400
@@ -1,4 +1,4 @@
-#SOURCES Copyrieht (c) 2000, 2011, Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 2000, 2011, Oracle and/or its affiliates. All rights reserved.
 # 
 # This program is free software; you can redistribute it and/or modify
 # it under the terms of the GNU General Public License as published by
@@ -51,13 +51,7 @@
 			rijndael.c my_aes.c sha1.c \
 			my_compare.c my_netware.c my_largepage.c \
 			my_memmem.c stacktrace.c \
-			my_windac.c my_access.c base64.c my_libwrap.c \
-		    thr_alarm.c thr_lock.c \
-			my_pthread.c my_thr_init.c \
-			thr_mutex.c thr_rwlock.c \
-			mf_soundex.c \
-			my_conio.c my_wincond.c my_winthread.c
-libmysys_a_CFLAGS = -fPIC
+			my_windac.c my_access.c base64.c my_libwrap.c
 
 if NEED_THREAD
 # mf_keycache is used only in the server, so it is safe to leave the file
@@ -66,8 +60,11 @@
 libmysys_a_SOURCES +=   mf_keycache.c
 endif
 
-EXTRA_DIST = CMakeLists.txt 
-#libmysys_a_LIBADD =	@THREAD_LOBJECTS@
+EXTRA_DIST =		thr_alarm.c thr_lock.c my_pthread.c my_thr_init.c \
+			thr_mutex.c thr_rwlock.c \
+			CMakeLists.txt mf_soundex.c \
+			my_conio.c my_wincond.c my_winthread.c
+libmysys_a_LIBADD =	@THREAD_LOBJECTS@
 # test_dir_DEPENDENCIES=	$(LIBRARIES)
 # testhash_DEPENDENCIES=	$(LIBRARIES)
 # test_charset_DEPENDENCIES=	$(LIBRARIES)
@@ -81,10 +78,10 @@
 			-DDEFAULT_SYSCONFDIR="\"$(sysconfdir)\"" \
                         @DEFS@
 
-#libmysys_a_DEPENDENCIES= @THREAD_LOBJECTS@
+libmysys_a_DEPENDENCIES= @THREAD_LOBJECTS@
 
 # I hope this always does the right thing. Otherwise this is only test programs
-FLAGS=$(DEFS) $(INCLUDES) $(CPPFLAGS) $(CFLAGS) @NOINST_LDFLAGS@ 
+FLAGS=$(DEFS) $(INCLUDES) $(CPPFLAGS) $(CFLAGS) @NOINST_LDFLAGS@
 
 #
 # The CP .. RM stuff is to avoid problems with some compilers (like alpha ccc)
--- server-tools/instance-manager/Makefile.am	2013-08-24 15:41:50.000000000 +0400
+++ ../../../stuff/experiments/mysql-5.1.65/server-tools/instance-manager/Makefile.am	2012-07-12 13:12:07.000000000 +0400
@@ -42,11 +42,11 @@
 # All symlinked files are grouped in libnet.a
 
 nodist_libnet_a_SOURCES= net_serv.cc client_settings.h
-libnet_a_LIBADD= $(top_builddir)/sql/libmysqldlib_la-password.$(OBJEXT) \
-	$(top_builddir)/sql/libmysqldlib_la-pack.$(OBJEXT) \
-	$(top_builddir)/sql/libmysqldlib_la-sql_state.$(OBJEXT) \
-	$(top_builddir)/sql/libmysqldlib_la-mini_client_errors.$(OBJEXT)\
-	$(top_builddir)/sql/libmysqldlib_la-client.$(OBJEXT)
+libnet_a_LIBADD= $(top_builddir)/sql/password.$(OBJEXT) \
+	$(top_builddir)/sql/pack.$(OBJEXT) \
+	$(top_builddir)/sql/sql_state.$(OBJEXT) \
+	$(top_builddir)/sql/mini_client_errors.$(OBJEXT)\
+	$(top_builddir)/sql/client.$(OBJEXT)
 
 CLEANFILES= net_serv.cc client_settings.h
